tests/common/utils.ts

```diff
@@ -0,0 +1,180 @@
+
+
+export interface ProofData {
+    proof: any;
+    publicSignals: any;
+    vk?: string;
+}
+
+export const loadProofData = (proofType: ProofType): ProofData => {
+    const dataPath = path.join(__dirname, 'data', `${proofType}.json`);
+    return JSON.parse(fs.readFileSync(dataPath, 'utf8'));
+};
+
+export const loadVerificationKey = (proofType: ProofType): string => {
+    if (proofType === ProofType.ultraplonk) {
+        const vkPath = path.join(__dirname, 'data', 'ultraplonk_vk.bin');
+        return fs.readFileSync(vkPath).toString('hex');
+    } else {
+        const dataPath = path.join(__dirname, 'data', `${proofType}.json`);
+        const proofData: ProofData = JSON.parse(fs.readFileSync(dataPath, 'utf8'));
+        return proofData.vk!;
+    }
+};
+
+export const validateEventResults = (eventResults: EventResults, expectAttestation: boolean): void => {
+    expect(eventResults.broadcastEmitted).toBe(true);
+    expect(eventResults.includedInBlockEmitted).toBe(true);
+    expect(eventResults.finalizedEmitted).toBe(true);
+    expect(eventResults.errorEventEmitted).toBe(false);
+
+    if (expectAttestation) {
+        expect(eventResults.attestationConfirmedEmitted).toBe(true);
+    } else {
+        expect(eventResults.attestationConfirmedEmitted).toBe(false);
+    }
+    expect(eventResults.attestationBeforeExpectedEmitted).toBe(false);
+    expect(eventResults.attestationMissedEmitted).toBe(false);
+};
+
+export const performVerifyTransaction = async (
+    proofType: ProofType,
+    proof: any,
+    publicSignals: any,
+    vk: string,
+    withAttestation: boolean,
+    validatePoe: boolean = false
+): Promise<{ eventResults: EventResults; transactionInfo: VerifyTransactionInfo }> => {
+    const session = await zkVerifySession.start().Testnet().withAccount(process.env.SEED_PHRASE!);
+
+    console.log(`${proofType} Executing transaction...`);
+    const verifier = session.verify()[proofType]();
+    const verify = withAttestation ? verifier.waitForPublishedAttestation() : verifier;
+
+    const { events, transactionResult } = await verify.execute(proof, publicSignals, vk);
+
+    console.log(`${proofType} Setting up event listeners...`);
+    const eventResults = withAttestation
+        ? handleEventsWithAttestation(events, proofType, 'verify')
+        : handleCommonEvents(events, proofType, 'verify');
+
+    console.log(`${proofType} Transaction result received. Validating...`);
+
+    const transactionInfo: VerifyTransactionInfo = await transactionResult;
+    validateVerifyTransactionInfo(transactionInfo, proofType, withAttestation);
+    validateEventResults(eventResults, withAttestation);
+
+    if (validatePoe) {
+        await validatePoE(session, transactionInfo.attestationId!, transactionInfo.leafDigest!);
+    }
+
+    console.log(`${proofType} Closing session...`);
+    await session.close();
+
+    return { eventResults, transactionInfo };
+};
+
+
+export const performVKRegistrationAndVerification = async (
```

Comment:
Perhaps it could be worth to also have another method for performing vk registration only ?

Reply:
These are just integration tests though that demonstrate the user flow. Adding a separate test to test the same thing again seems unnecessary?